### PR TITLE
feat(ec2): add aws_vpc_security_group_ingress/egress_rule resource

### LINKED ISSUE #683: Request for `aws_vpc_security_group_egress_rule` and `aws_vpc_security_group_ingress_rule` resource
### What resource do you need?

Terraform Resource Names: 
- `aws_vpc_security_group_egress_rule`
- `aws_vpc_security_group_ingress_rule`

### What is your use case?

These two resources are the replacements for `aws_security_group_rule`. I have encountered situations where Terraform (and by extension AWS Provider) cannot reconcile changes with `aws_security_group_rule` resources. There is also a warning front and center on [the resource docs](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/security_group_rule) about potential issues with the old security group rule.

### PR DESCRIPTION
**Pull request by huynhsontung**
### Description of your changes

Fixes #683

Generated code for `aws_vpc_security_group_ingress_rule` and `aws_vpc_security_group_egress_rule`.

I have:

- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

Tested in a personal cluster

```yaml
apiVersion: ec2.aws.upbound.io/v1beta1
kind: VPCSecurityGroupIngressRule
metadata:
  annotations:
    crossplane.io/external-create-succeeded: '2023-09-27T17:01:35-07:00'
    crossplane.io/external-name: sgr-09ce7f826ed2b32c8
    meta.upbound.io/example-id: ec2/v1beta1/vpcsecuritygroupingressrule
    upjet.crossplane.io/provider-meta: ''
  creationTimestamp: '2023-09-28T00:00:42Z'
  finalizers:
    - finalizer.managedresource.crossplane.io
  generation: 4
  name: ingress-sgr-test
  resourceVersion: '86044428'
  uid: 8efe5fe6-d2b7-4879-9fce-d798f97abeb0
spec:
  deletionPolicy: Delete
  forProvider:
    cidrIpv4: 10.22.10.0/24
    fromPort: 10250
    ipProtocol: tcp
    region: us-east-2
    securityGroupId: sg-0e0c05b477edebd15
    tags:
      crossplane-kind: vpcsecuritygroupingressrule.ec2.aws.upbound.io
      crossplane-name: ingress-sgr-test
      crossplane-providerconfig: default
    toPort: 10250
  initProvider: {}
  managementPolicies:
    - '*'
  providerConfigRef:
    name: default
status:
  atProvider:
    arn: >-
      arn:aws:ec2:us-east-2:******:security-group-rule/sgr-09ce7f826ed2b32c8
    cidrIpv4: 10.22.10.0/24
    fromPort: 10250
    id: sgr-09ce7f826ed2b32c8
    ipProtocol: tcp
    securityGroupId: sg-0e0c05b477edebd15
    securityGroupRuleId: sgr-09ce7f826ed2b32c8
    tags:
      crossplane-kind: vpcsecuritygroupingressrule.ec2.aws.upbound.io
      crossplane-name: ingress-sgr-test
      crossplane-providerconfig: default
    tagsAll:
      crossplane-kind: vpcsecuritygroupingressrule.ec2.aws.upbound.io
      crossplane-name: ingress-sgr-test
      crossplane-providerconfig: default
    toPort: 10250
  conditions:
    - lastTransitionTime: '2023-09-28T00:01:35Z'
      reason: ReconcileSuccess
      status: 'True'
      type: Synced
    - lastTransitionTime: '2023-09-28T00:01:50Z'
      reason: Available
      status: 'True'
      type: Ready
    - lastTransitionTime: '2023-09-28T00:01:38Z'
      reason: Success
      status: 'True'
      type: LastAsyncOperation
    - lastTransitionTime: '2023-09-28T00:01:38Z'
      reason: Finished
      status: 'True'
      type: AsyncOperation
```

### CODE REVIEW

**Comment by jeanduplessis** [comment]
/test-examples="examples/vpc/securitygroupegressrule.yaml,examples/vpc/securitygroupingressrule.yaml"

File: examples/vpc/securitygroupegressrule.yaml [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    cidrIpv4: 10.0.0.0/8
+    fromPort: 80
+    ipProtocol: tcp
+    region: us-west-1
+    securityGroupIdSelector:
+      matchLabels:
+        testing.upbound.io/example-name: example
+    toPort: 8080
+
+---
```
Comment (jeanduplessis) [inline examples/vpc/securitygroupegressrule.yaml:20]: @huynhsontung please remove this trailing object separator

```suggestion
```

File: examples/vpc/securitygroupingressrule.yaml [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    cidrIpv4: 10.0.0.0/8
+    fromPort: 80
+    ipProtocol: tcp
+    region: us-west-1
+    securityGroupIdSelector:
+      matchLabels:
+        testing.upbound.io/example-name: example
+    toPort: 8080
+
+---
```
Comment (jeanduplessis) [inline examples/vpc/securitygroupingressrule.yaml:20]: Also this one please
```suggestion
```

**Review by jeanduplessis — CHANGES_REQUESTED** [review]
@huynhsontung, thanks for the contribution. Two small changes please to clean up the examples.

**Comment by huynhsontung** [comment]
/test-examples="examples/vpc/securitygroupegressrule.yaml,examples/vpc/securitygroupingressrule.yaml"

**Comment by huynhsontung** [comment]
```
case.go:363: retrieving API resource for vpc.aws.upbound.io/v1beta1, Kind=SecurityGroupEgressRule failed: the server could not find the requested resource
case.go:363: retrieving API resource for vpc.aws.upbound.io/v1beta1, Kind=SecurityGroupIngressRule failed: the server could not find the requested resource
```

The package was built and deployed successfully but could not find the API resource. I'm not sure what is missing.

**Comment by huynhsontung** [comment]
@turkenf Sorry for the late followup. I went back and updated some missing codegen configs to reference external resources correctly. However, I still get this error message regardless of what config I change.

```
cannot run refresh: refresh failed: reading VPC Security Group Rule (): InvalidSecurityGroupRuleId.Malformed: The security group rule ID '' is malformed
        status code: 400, request id: bb0840ad-3c70-4405-8c03-ffd864b68399      {"type": "Warning", "object": {"kind":"SecurityGroupEgressRule","name":"egressrule-test","uid":"1f43bf65-2d6d-4b97-af92-4e2d55b6f815","apiVersion":"vpc.aws.upbound.io/v1beta1","resourceVersion":"18950285"}, "reason": "CannotObserveExternalResource"}
```

terraform.tfstate
```json
{
  "version": 4,
  "terraform_version": "1.2.1",
  "serial": 1,
  "lineage": "1f43bf65-2d6d-4b97-af92-4e2d55b6f815",
  "outputs": null,
  "resources": [
    {
      "mode": "managed",
      "type": "aws_vpc_security_group_egress_rule",
      "name": "egressrule-test",
      "provider": "provider[\"registry.terraform.io/hashicorp/aws\"]",
      "instances": [
        {
          "schema_version": 0,
          "attributes": {
            "cidr_ipv4": "10.0.0.0/8",
            "from_port": 80,
            "id": "", 
            "ip_protocol": "tcp",
            "security_group_id": "sg-0123456789abc",
            "tags": {
              "crossplane-kind": "securitygroupegressrule.vpc.aws.upbound.io",
              "crossplane-name": "egressrule-test",
              "crossplane-providerconfig": "default"
            },
            "to_port": 8080
          }
        }
      ]
    }
  ]
}
```

It seems that on the first reconciliation, the provider generates a stub tfstate file with an empty `id` field and Terraform doesn't like that. If I inject the SGR external name using `crossplane.io/external-name` annotation then it starts to reconcile just fine.

I am not aware of any way to customize the provider tfstate generation behaviour and am not sure where to go from here. I would appreciate some guidance on this issue.

**Comment by turkenf** [comment]
My observations in this PR with provider-aws v0.36.0:

I keep getting the same error:
```
      message: "observe failed: cannot run refresh: refresh failed: reading VPC Security
      Group Rule (): InvalidSecurityGroupRuleId.Malformed: The security group rule
      ID '' is malformed\n\tstatus code: 400, request id: 39c253e2-4845-4349-8774-62a1969a27b9"
```
When I create the `egress_rule` resource in terraform with the example below:
```
resource "aws_vpc_security_group_egress_rule" "example" {
  security_group_id = "sg-0e625fc4bf678386f"

  cidr_ipv4   = "10.0.0.0/8"
  from_port   = 80
  ip_protocol = "tcp"
  to_port     = 8080
}
```
The attributes.id in the tfstate file comes as `"id": "sgr-0ec075aef37950e08",`:
```
          "schema_version": 0,
          "attributes": {
            "arn": "arn:aws:ec2:us-west-1:569674547146:security-group-rule/sgr-0ec075aef37950e08",
            "cidr_ipv4": "10.0.0.0/8",
            "cidr_ipv6": null,
            "description": null,
            "from_port": 80,
            "id": "sgr-0ec075aef37950e08",
            "ip_protocol": "tcp",
            "prefix_list_id": null,
            "referenced_security_group_id": null,
            "security_group_id": "sg-0e625fc4bf678386f",
            "security_group_rule_id": "sgr-0ec075aef37950e08",
            "tags": null,
            "tags_all": {},
            "to_port": 8080
          },
          "sensitive_attributes": []
```
In our example, the `id` in the `tfstate` file is somehow empty and this causes an error and the resource is not created:
```
                    "schema_version": 0,
                    "attributes": {
                        "cidr_ipv4": "10.0.0.0/8",
                        "from_port": 80,
                        "id": "",
                        "ip_protocol": "tcp",
                        "security_group_id": "sg-0e625fc4bf678386f",
                        "tags": {
                            "crossplane-kind": "securitygroupegressrule.vpc.aws.upbound.io",
                            "crossplane-name": "example",
                            "crossplane-providerconfig": "default"
                        },
                        "to_port": 8080
```

**Comment by jeanduplessis** [comment]
/test-examples="examples/vpc/securitygroupegressrule.yaml,examples/vpc/securitygroupingressrule.yaml"

File: config/vpc/config.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+*/
+
+package vpc
+
+import (
+	"github.com/upbound/upjet/pkg/config"
+)
+
+// Configure adds configurations for vpc group.
+func Configure(p *config.Provider) {
```
Comment (ulucinar) [inline config/vpc/config.go:12]: Hi @turkenf,
We had better understand whether we still need the `SecurityGroupRule.ec2` resource as two separate ingress and egress rules are being introduced. Looking at https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/security_group_rule, Terraform has not deprecated this resource but we may still mark it as deprecated if there are not configurations not covered by these new two resources (for instance, are there any other rule types other than ingress and egress, etc.)?

**Comment by huynhsontung** [comment]
@turkenf @ulucinar I have found a solution for the malformed ID issue. Terraform performs state validation so when the rule's ID is empty, it raises an error and exits. I found out that by providing a fake ID value that has `sgr-` prefix (I used `sgr-stub`), it was able to pass validation and reconcile successfully. Once the provider gets past the initial resource creation, the proper SGR ID is populated in the state file and the external name as expected. Tested successfully in my own cluster.

I also moved the resources into the `ec2` group.

**Comment by turkenf** [comment]
/test-examples="examples/vpc/securitygroupegressrule.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/vpc/securitygroupingressrule.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/ec2/vpcsecuritygroupegressrule.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/ec2/vpcsecuritygroupingressrule.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/ec2/vpcsecuritygroupingressrule.yaml"

File: config/ec2/config.go [hunk 4]
```diff
@@ -217,6 +217,30 @@ func Configure(p *config.Provider) {
 		}
 	})
 
+	p.AddResourceConfigurator("aws_vpc_security_group_ingress_rule", func(r *config.Resource) {
```
[the remark this answers is not in the document]
Comment (huynhsontung) [reply config/ec2/config.go:220]: Totally agree.

File: config/ec2/config.go [hunk 5]
```diff
@@ -217,6 +217,30 @@ func Configure(p *config.Provider) {
...
+		r.References["security_group_id"] = config.Reference{
+			Type: "github.com/upbound/provider-aws/apis/ec2/v1beta1.SecurityGroup",
```
[the remark this answers is not in the document]
Comment (huynhsontung) [reply config/ec2/config.go]: I missed updating these references when moving the resources from the `vpc` to `ec2` group. I think we can use `Type: "SecurityGroup"` now that they are all in the same `ec2` group, similar to other configs in this file.

**Comment by turkenf** [comment]
/test-examples="examples/ec2/vpcsecuritygroupingressrule.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/ec2/securitygroupingressrule.yaml"

**Comment by turkenf** [comment]
/test-examples="examples/ec2/securitygroupegressrule.yaml"

**Comment by huynhsontung** [comment]
Apply is failing after rebase. I will debug this on my end. Not sure why import didn't pass.

**Comment by huynhsontung** [comment]
@turkenf To speed up the debug process, would you mind rerunning the uptest on one of the resources?

**Comment by turkenf** [comment]
> Apply is failing after rebase. I will debug this on my end. Not sure why import didn't pass.

With the latest uptest update, update and import tests were added to automatic testing, these were not tested before. Have you done the import test of the resources locally?

**Comment by turkenf** [comment]
/test-examples="examples/ec2/securitygroupingressrule.yaml"

**Comment by huynhsontung** [comment]
> With the latest uptest update, update and import tests were added to automatic testing, these were not tested before. Have you done the import test of the resources locally?

I tested import before rebase. I can see from the log that the `atProvider` object is populated. The test failed when [trying to check the VPC ID](https://github.com/upbound/provider-aws/actions/runs/6466512888/job/17554755330#step:10:5130).

**Comment by turkenf** [comment]
@huynhsontung, could you please rebase again after this [PR](https://github.com/upbound/provider-aws/pull/914) merged and revert the external name change?

File: config/ec2/config.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (turkenf) [reply config/ec2/config.go]: It can stay this way, but the safest way to set references is to use `TerraformName`.

**Comment by turkenf** [comment]
/test-examples="examples/ec2/securitygroupingressrule.yaml"

**Review by turkenf — APPROVED** [review]
Thank you so much for your effort in this PR @huynhsontung, LGTM.

**Review by jeanduplessis — APPROVED** [review]

**Pull request merged**